File: cpp/StrangeCharacters/src/Character.cpp
#include "Character.hpp"

namespace Characters {

Character::Character(std::string firstName, std::optional<std::string> lastName, bool isMonster)
: FirstName(firstName)
, LastName(lastName)
, IsMonster(isMonster)
{
}

void Character::AddChild(Character* child)
{
    children.insert(child);
    child->parents.insert(this);
    for (auto character : children)
    {
        if (child->FirstName != character->FirstName)
        {
            child->siblings.insert(character);
            character->siblings.insert(child);
        }
    }
}

std::string to_string(const Character& character)
{
    return "Character(" + character.FirstName + ", " + (character.LastName ? *character.LastName : "null") + ", isMonster:" + std::to_string(character.IsMonster) + ")";
}

bool Character::operator==(const Character& other) const
{
    return FirstName == other.FirstName;
}

}

File: java/src/main/java/org/example/Character.java
package org.example;

import java.util.HashSet;

public class Character {
  public final String firstName;
  public final String lastName;
  public final boolean isMonster;

  private Character nemesis;

  public HashSet<Character> children = new HashSet<>();
  public HashSet<Character> parents = new HashSet<>();
  public HashSet<Character> siblings = new HashSet<>();

  public Character(String firstName, String lastName, boolean isMonster) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.isMonster = isMonster;
  }

  public Character(String firstName) {
    this(firstName, null, false);
  }

  public Character(String firstName, boolean isMonster) {
    this(firstName, null, isMonster);
  }

  public Character(String firstName, String lastName) {
    this(firstName, lastName, false);
  }

  public void addChild(Character child) {
    children.add(child);
    child.parents.add(this);
    for (Character character : children) {
      if (!child.firstName.equals(character.firstName)) {
        child.siblings.add(character);
        character.siblings.add(child);
      }
    }
  }

  public Character getNemesis() {
    return nemesis;
  }

  void setNemesis(Character monster) {
    this.nemesis = monster;
  }

  @Override
  public String toString() {
    return String.format("Character(%s, %s, %s)", firstName, lastName, isMonster ? "True" : "False");
  }

  protected boolean equals(Character other) {
    return firstName.equals(other.firstName);
  }

  @Override
  public boolean equals(Object o) {
    if (null == o) return false;
    if (this == o) return true;
    if (o.getClass() != this.getClass()) return false;
    return equals((Character) o);
  }

  @Override
  public int hashCode() {
    return firstName.hashCode();
  }
}
